### ISSUE TITLE
Resnet56 Training Results, compared to another Tensorflow 2 Model

### ISSUE THREAD

**Issue by drcdr** [lambdal/TensorFlow2-tutorial#1]
I'm trying to learn TF2+Keras, and came across your great examples.  I'm working my way through the image classification example.  I'm comparing this to what I think is the official TF 2.0 Keras-based approach(?), [here](https://github.com/tensorflow/models/blob/master/official/vision/image_classification/resnet_cifar_main.py), which I'll call MOVIR (after the first initials in the path of the python file, within tensorflow/models).

When I run MOVIR (vanilla, no command line arguments), I get the following training & validation results (Val Acc=93.22%):
```
390/390 - 19s - loss: 0.1213 - categorical_accuracy: 0.9995
78/78 - 2s - loss: 0.4276 - categorical_accuracy: 0.9322
```
I've modified your example for resnet_cifar.py, changing #GPUs=1 and the # of epochs and the LR schedule to match MOVIR:
```
182, [(0.1, 91), (0.01, 136), (0.001, 182)]
```
The final results I get using the modified resnet_cifar.py are (Val Acc=78.32%):
```
Epoch 182/182
390/390 [==============================] - 30s 78ms/step - loss: 0.1410 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.1042 - val_sparse_categorical_accuracy: 0.7832
78/78 [==============================] - 1s 17ms/step - loss: 1.1042 - sparse_categorical_accuracy: 0.7832
```
Both are using Resnet56 and BS=128.
[I'm using Windows10, TitanXP, Python 3.7, tf-nightly-gpu-2.1.0.dev20191028.]

_Questions_
* Do you get similar results?  If so, any idea on why the val acc results are low?
* Thought: I don't see how MOVIR does augmentation, so I don't see how they reach a relatively decent result.

_Additional Information_
I noticed that the Resnet56 model that MOVIR uses is slightly different than the one that you use.  So, I thought I'd try switching them.  Here is the summary (LL = Lambda Labs).  So, I think the performance difference is not due to the slightly different models.

Training   File | Resnet56   Model | Final   Validation Accuracy | ~s/Epoch
-- | -- | -- | --
LL | LL | 78.32% | 31
LL | MOVIR | 78.52% | 27
MOVIR | LL | 93.46%  | 21
MOVIR | MOVIR | 93.22% | 19

**Comment by drcdr** [comment]
@chuanli11 or @stephenbalaban - any thoughts?

**Comment by bozcani** [comment]
I see that Resnet56 model in LL includes bias term and regularizers for the convolutional layers. However, most of ResNet implementation does not include bias term for convolutional layers (like in MOVIR) since batch normalization layers already include bias term. These discussions might be helpful:
https://github.com/KaimingHe/deep-residual-networks/issues/10
https://stackoverflow.com/questions/45134831/is-bias-necessarily-need-at-colvolution-layer
https://stackoverflow.com/questions/46256747/can-not-use-both-bias-and-batch-normalization-in-convolution-layers

Could you please remove bias terms in convolutional layers and run the experiment again? Not sure this is the real reason but I hope it can help you.

**Comment by bozcani** [comment]
Btw, I updated resnet.py in my fork. You can check it if you are interested.
https://github.com/bozcani/TensorFlow2-tutorial

**Comment by drcdr** [comment]
@bozcani Thank you for the feedback.  I tried your fork, changing it just slightly to run the scenario above.  I got results similar to above (run1=79.33%; run2=78.7%).  If you have time, do you mind trying to see what you get, to make sure it's not machine/environment variations?  I simply added the following lines after the other upper-case defines at the top of resnet_cifar.py, to overwrite their previous values:
```
NUM_GPUS = 1
NUM_EPOCHS, LR_SCHEDULE = 182, [(0.1, 91), (0.01, 136), (0.001, 182)]
```
Meanwhile, I've also tried changing the training and validation labels to categorical for the LL implementation, and using categorical loss/metrics.  I got 79.15%.

From above, the poor performance is following the LL harness, not the (interchanged) model.  So, I'm tempted to just dump all of the data going into the model.fit function for the two scenarios, and see what's different.  That's why I tried moving to a common categorical approach, first.

In both LL and MOVIR, the training loss ends at about 0.12, 0.13.  For MOVIR, the Val Loss ends up at about 0.42.  For LL, the Val Loss ends up at ~1.1.  So...maybe a label problem or something?  I've also seen this where the validation dataloader is running training-like augmentations, but that doesn't seem to be the case here.

**Comment by bozcani** [comment]
@drcdr Good points! I will try and get you back, hopefully, tomorrow.

**Comment by drcdr** [comment]
Converting the CIFAR10 image data to float32 seems to be a necessary but not sufficient thing to do.  I probably could have converted one-time up-front, but I added an initial map for the train and test datasets:
```
def tofloat(x, y):
  x, y = tf.cast(x, tf.float32), tf.cast(y, tf.int32)
  return x, y
```
It improves the ```182, [(0.1, 91), (0.01, 136), (0.001, 182)]``` schedule from ~78% to ~88%, but does not not get to ~93%.   I thought I might try to take some shortcuts and run the ```60, [(0.1, 30), (0.01, 45)]``` schedule, but ... it would appear that the problem does not show up there.

**Comment by drcdr** [comment]
After more careful testing, it looks like the ```tofloat``` modification for LL is enough to fix the problem, so that LL results match MOVIR results.

The figure below shows the validation accuracy results, from several different testing trials.
- Green is MOVIR (with the MOVIR model, use_bias=false); Red is LL (with the LL model, with bias)
- Runs cover {Resnet20, Resnet32, Resnet56} x {LL, MOVIR} x {2 schedules}, where the schedules are the ones mentioned above.  The Resnet20-MOVIR cases are repeated.
- The 'x' marks the best result, not the end result.  The number in the tag refers to the accuracy at the last epoch.
- the figure is zoomed in to show that for each case, the LL validation accuracy and MOVIR validation accuracy compare 'close enough'.

So, as far as I'm concerned, this issue can be closed.  To train CIFAR10 to higher accuracy with this example, incorporate the ```to_float()``` fix (and add to train_dataset and test_dataset) or something equivalent.

Also, I modified LL to use categorical labels
```
y, y_test = tf.keras.utils.to_categorical(y), tf.keras.utils.to_categorical(y_test)
```
and loss, and metrics.

Lastly, I will mention that I found that MOVIR is 1.7x to 2.2x faster than LL (TitanXP).  I'm guessing it's because of [this function](https://github.com/tensorflow/models/blob/master/official/vision/image_classification/common.py#L284), but I don't know.

![LL_vs_MOVIR](https://user-images.githubusercontent.com/18181131/68559907-adc97080-040c-11ea-9ce6-be8caa2713f2.png)
